main.go

```diff
@@ -0,0 +1,297 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+	http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// The main package for the executable
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/imdario/mergo"
+	"gopkg.in/yaml.v3"
+
+	"github.com/KohlsTechnology/hierarchy/pkg/version"
+	"github.com/kylelemons/godebug/diff"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type config struct {
+	hierarchyFile   string
+	basePath        string
+	outputFile      string
+	filterExtension string
+	printVersion    bool
+	diffOutput      bool
+	logDebug        bool
+	logTrace        bool
+	failMissing     bool
+}
+
+// Default file filter
+const defaultFileFilter = "(.yaml|.yml|.json)$"
+
+func parseFlags() config {
+	application := kingpin.New(filepath.Base(os.Args[0]), "Hierarchy")
+	application.HelpFlag.Short('h')
+
+	cfg := config{}
+
+	application.Flag("file", "Path and name of the hierarchy file.").Short('f').
+		Envar("HIERARCHY_FILE").Default("./hierarchy.lst").StringVar(&cfg.hierarchyFile)
+	application.Flag("output", "Path and name of the output file.").Short('o').
+		Envar("HIERARCHY_OUTPUT").Default("./output.yaml").StringVar(&cfg.outputFile)
+	application.Flag("filter", "Regex for file extension of files being merged").Short('i').
+		Envar("HIERARCHY_FILTER").Default(defaultFileFilter).StringVar(&cfg.filterExtension)
+	application.Flag("trace", "Prints a diff after processing each file. This generates A LOT of output").
+		Envar("HIERARCHY_TRACE").Default("false").BoolVar(&cfg.logTrace)
+	application.Flag("failmissing", "Fail if a directory in the hierarchy is missing").Short('m').
+		Envar("HIERARCHY_FAILMISSING").Default("false").BoolVar(&cfg.failMissing)
+	application.Flag("version", "Print version and build information, then exit").Short('V').
+		Default("false").BoolVar(&cfg.printVersion)
+	application.Flag("debug", "Print debug output").Short('d').
+		Envar("HIERARCHY_DEBUG").Default("false").BoolVar(&cfg.logDebug)
+
+	_, err := application.Parse(os.Args[1:])
+
+	if cfg.printVersion {
+		version.Print()
+		os.Exit(0)
+	}
+
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrapf(err, "Error parsing commandline arguments"))
+		application.Usage(os.Args[1:])
+		os.Exit(2)
+	}
+	return cfg
+}
+
+// checkForError fails the program with a fatal error message if e != nil
+func checkForError(e error) {
+	if e != nil {
+		log.Fatal(e)
+	}
+}
+
+func main() {
```

Comment:
Would it make sense to move the main() function to the very bottom? It doesn't affect anything, but it's kind of weird to see it in the middle mixed with other functions.